File: packages/coterie-js/src/nodes/bracket.ts
import type { LLMClient } from "../core/llm/base.js";
import type { CoterieState } from "../core/state.js";

const BRACKET_JUDGE_SYSTEM = `You are an impartial bracket judge for an N-way coding tournament.
You see N attempts at the same task, one per agent. Score each on the configured criteria,
then rank from best to worst.

Return strict JSON only — no prose, no markdown:
{
  "ranking": [{"agent_id": "...", "score": <int 1-100>, "reason": "<one sentence>"}, ...],
  "winner": "<agent_id>",
  "summary": "<2-3 sentences>"
}`;

async function rankCurrentRound(
  llm: LLMClient | null,
  task: string,
  attempts: any[],
  criteria: string[],
): Promise<{ ranking: any[]; winner: string; summary: string }> {
  if (!llm) {
    const successful = attempts.filter((a) => a.exit_code === 0);
    const pool = successful.length ? successful : attempts;
    const chosen = pool.reduce((a, b) =>
      (a.cost_estimate_usd ?? Infinity) <= (b.cost_estimate_usd ?? Infinity) ? a : b,
    );
    return { ranking: [], winner: chosen.agent_id, summary: "no judge LLM; chose cheapest" };
  }

  const block = attempts
    .map((a) =>
      `--- ${a.agent_id} ---\nexit_code: ${a.exit_code}, duration_s: ${a.duration_s.toFixed(2)}, cost: $${a.cost_estimate_usd}\n` +
      `stdout (first 1500 chars):\n${a.stdout.slice(0, 1500)}`,
    )
    .join("\n\n");
  const prompt = `Task: ${task}\n\nCriteria (priority order): ${JSON.stringify(criteria)}\n\nAttempts:\n${block}`;
  const raw = await llm.chat(BRACKET_JUDGE_SYSTEM, [{ role: "user", content: prompt }]);
  try {
    const d = JSON.parse(raw);
    return { ranking: d.ranking ?? [], winner: d.winner, summary: d.summary ?? "" };
  } catch {
    const successful = attempts.filter((a) => a.exit_code === 0);
    const pool = successful.length ? successful : attempts;
    const chosen = pool.reduce((a, b) =>
      (a.cost_estimate_usd ?? Infinity) <= (b.cost_estimate_usd ?? Infinity) ? a : b,
    );
    return { ranking: [], winner: chosen.agent_id, summary: `bracket judge unparseable: ${raw.slice(0, 120)}` };
  }
}

export function makeBracketJudgeNode(llm: LLMClient | null) {
  return async (state: CoterieState) => {
    const cfg = state.config;
    const tournament = cfg.tournament ?? {};
    const judgeCfg = tournament.judge ?? {};
    const criteria = judgeCfg.criteria ?? ["correctness", "minimal-diff", "tests-pass", "clarity"];
    const participantsAll = (tournament.participants ?? []) as string[];
    const totalRounds = tournament.rounds ?? 1;

    const modeState = { ...(state.mode_state ?? {}) };
    let roundIdx = (modeState.tournament_round_idx ?? 0) as number;
    const eliminated = [...(modeState.eliminated_participants ?? [])] as string[];
    const lastCount = (modeState.last_judged_run_count ?? 0) as number;

    const allRuns = state.runs ?? [];
    const thisRound = allRuns
      .slice(lastCount)
      .filter((r) => r.role === "tournament-participant" && participantsAll.includes(r.agent_id) && !eliminated.includes(r.agent_id));

    if (thisRound.length === 0) return { status: "failed" as const };

    const { ranking, winner, summary } = await rankCurrentRound(llm, state.task, thisRound, criteria);
    roundIdx += 1;

    let survivors: string[];
    let newEliminated: string[];
    if (totalRounds === 1 || thisRound.length <= 2) {
      survivors = [winner];
      newEliminated = [...eliminated, ...thisRound.filter((a) => a.agent_id !== winner).map((a) => a.agent_id)];
    } else {
      const survivorsN = Math.max(1, Math.floor(thisRound.length / 2));
      const ordered = ranking.length
        ? ranking.map((r) => r.agent_id).filter(Boolean)
        : [winner, ...thisRound.filter((a) => a.agent_id !== winner).map((a) => a.agent_id)];
      survivors = ordered.slice(0, survivorsN);
      newEliminated = [...eliminated, ...ordered.filter((a) => !survivors.includes(a))];
    }

    modeState.tournament_round_idx = roundIdx;
    modeState.eliminated_participants = newEliminated;
    modeState.survivors = survivors;
    modeState.last_judged_run_count = allRuns.length;
    modeState.bracket_ranking = ranking;
    modeState.bracket_summary = summary;
    const done = survivors.length <= 1 || roundIdx >= totalRounds;
    if (done) modeState.winner = survivors[0] ?? winner;

    return {
      mode_state: modeState,
      judge_history: [{
        step: state.current_step_idx ?? 0,
        winner,
        reason: summary,
        scores: Object.fromEntries(ranking.map((r) => [r.agent_id, r.score ?? 0])),
      }],
      status: done ? ("done" as const) : ("tournament_round" as const),
    };
  };
}

File: packages/coterie-py/src/coterie/nodes/bracket.py
"""Bracket judge for `tournament` mode.

Two modes:
- `rounds == 1` (default): single N-way ranking. Picks one winner outright.
- `rounds > 1`: bracket elimination. Each round halves the field (rounded down,
  minimum 1). Loops via a state-driven conditional edge in the tournament graph.

State management uses `mode_state`:
- `tournament_round_idx` — incremented each judge call
- `eliminated_participants` — accumulating list of agent_ids that lost their round
- `survivors` — last round's survivors (used to wire the next round's prompts)
- `last_judged_run_count` — index into `runs` so the judge only scores the
  current round, not all historical runs

The judge's LLM call is identical in both modes; only the routing logic differs.
"""

import json
from typing import Any

from coterie.core.llm.base import LLMClient
from coterie.core.state import CoterieState

BRACKET_JUDGE_SYSTEM = """You are an impartial bracket judge for an N-way coding tournament.
You see N attempts at the same task, one per agent. Score each on the configured criteria,
then rank from best to worst.

Return strict JSON only — no prose, no markdown:
{
  "ranking": [{"agent_id": "...", "score": <int 1-100>, "reason": "<one sentence>"}, ...],
  "winner": "<agent_id>",
  "summary": "<2-3 sentences>"
}
"""


def _rank_current_round(
    llm: LLMClient | None,
    task: str,
    attempts: list[dict],
    criteria: list[str],
) -> dict:
    """Returns dict with ranking, winner, summary. Falls back to cheapest-successful when no LLM."""
    if not llm:
        successful = [a for a in attempts if a["exit_code"] == 0]
        chosen = min(
            successful or attempts,
            key=lambda a: a.get("cost_estimate_usd") or float("inf"),
        )
        return {
            "ranking": [],
            "winner": chosen["agent_id"],
            "summary": "no judge LLM configured; chose cheapest successful attempt",
        }

    attempts_block = "\n\n".join(
        f"--- {a['agent_id']} ---\n"
        f"exit_code: {a['exit_code']}, duration_s: {a['duration_s']:.2f}, "
        f"cost: ${a.get('cost_estimate_usd')}\n"
        f"stdout (first 1500 chars):\n{a['stdout'][:1500]}"
        for a in attempts
    )
    prompt = (
        f"Task: {task}\n\n"
        f"Criteria (priority order): {criteria}\n\n"
        f"Attempts:\n{attempts_block}"
    )
    raw = llm.chat(BRACKET_JUDGE_SYSTEM, [{"role": "user", "content": prompt}])
    try:
        decision = json.loads(raw)
        return {
            "ranking": decision.get("ranking", []),
            "winner": decision["winner"],
            "summary": decision.get("summary", ""),
        }
    except (json.JSONDecodeError, KeyError):
        successful = [a for a in attempts if a["exit_code"] == 0]
        chosen = min(
            successful or attempts,
            key=lambda a: a.get("cost_estimate_usd") or float("inf"),
        )
        return {
            "ranking": [],
            "winner": chosen["agent_id"],
            "summary": f"bracket judge output unparseable: {raw[:120]!r}",
        }


def make_bracket_judge_node(llm: LLMClient | None = None):
    def judge(state: CoterieState) -> dict[str, Any]:
        cfg = state["config"]
        tournament = cfg.get("tournament") or {}
        judge_cfg = tournament.get("judge") or {}
        criteria = judge_cfg.get("criteria") or ["correctness", "minimal-diff", "tests-pass", "clarity"]
        participants_all = tournament.get("participants") or []
        total_rounds = tournament.get("rounds", 1)

        mode_state = dict(state.get("mode_state") or {})
        round_idx = mode_state.get("tournament_round_idx", 0)
        eliminated = list(mode_state.get("eliminated_participants") or [])
        last_count = mode_state.get("last_judged_run_count", 0)

        # Score only this round's runs.
        all_runs = state.get("runs") or []
        this_round = [
            r
            for r in all_runs[last_count:]
            if r.get("role") == "tournament-participant"
            and r["agent_id"] in participants_all
            and r["agent_id"] not in eliminated
        ]

        if not this_round:
            return {"status": "failed"}

        decision = _rank_current_round(llm, state["task"], this_round, criteria)
        winner_id = decision["winner"]
        ranking = decision["ranking"]
        summary = decision["summary"]

        # Determine survivors. Single-round (rounds=1): winner only. Multi-round:
        # top half (rounded down, minimum 1).
        round_idx += 1
        if total_rounds == 1 or len(this_round) <= 2:
            survivors = [winner_id]
            new_eliminated = eliminated + [a["agent_id"] for a in this_round if a["agent_id"] != winner_id]
        else:
            survivors_n = max(1, len(this_round) // 2)
            # Use ranking if available; else fall back to {winner first, rest in input order}.
            if ranking:
                ordered = [r["agent_id"] for r in ranking if r.get("agent_id")]
            else:
                ordered = [winner_id] + [a["agent_id"] for a in this_round if a["agent_id"] != winner_id]
            survivors = ordered[:survivors_n]
            new_eliminated = eliminated + [a for a in ordered if a not in survivors]

        mode_state["tournament_round_idx"] = round_idx
        mode_state["eliminated_participants"] = new_eliminated
        mode_state["survivors"] = survivors
        mode_state["last_judged_run_count"] = len(all_runs)
        mode_state["bracket_ranking"] = ranking
        mode_state["bracket_summary"] = summary

        # Terminate when one survivor remains or we've consumed all rounds.
        done = len(survivors) <= 1 or round_idx >= total_rounds
        if done:
            mode_state["winner"] = survivors[0] if survivors else winner_id

        return {
            "mode_state": mode_state,
            "judge_history": [
                {
                    "step": state.get("current_step_idx", 0),
                    "winner": winner_id,
                    "reason": summary,
                    "scores": {
                        entry.get("agent_id", ""): entry.get("score", 0) for entry in ranking
                    },
                }
            ],
            "status": "done" if done else "tournament_round",
        }

    return judge
